services/private_ip_manager/src/test/java/com/futurewei/alcor/privateipmanager/repo/IpAddrTest.java

```diff
@@ -0,0 +1,180 @@
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+package com.futurewei.alcor.privateipmanager.repo;
+
+import com.futurewei.alcor.common.db.ICache;
+import com.futurewei.alcor.common.db.MockCache;
+import com.futurewei.alcor.common.db.Transaction;
+import com.futurewei.alcor.privateipmanager.config.UnitTestConfig;
+import com.futurewei.alcor.privateipmanager.entity.IpAddrAlloc;
+import com.futurewei.alcor.privateipmanager.entity.IpAddrRange;
+import com.futurewei.alcor.privateipmanager.entity.VpcIpRange;
+import com.futurewei.alcor.privateipmanager.exception.IpAddrAllocNotFoundException;
+import com.futurewei.alcor.privateipmanager.exception.IpAddrNotEnoughException;
+import com.futurewei.alcor.privateipmanager.exception.IpRangeNotFoundException;
+import com.futurewei.alcor.privateipmanager.exception.NotFoundIpRangeFromVpc;
+import com.futurewei.alcor.web.entity.ip.IpAddrRangeRequest;
+import com.futurewei.alcor.web.entity.ip.IpAddrRequest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static com.futurewei.alcor.privateipmanager.util.IpAddressBuilder.buildIpAddrRangeRequest;
+import static com.futurewei.alcor.privateipmanager.util.IpAddressBuilder.buildIpAddrRequest;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.*;
+
+public class IpAddrTest {
+    private IpAddrRangeRepo ipAddrRangeRepo;
+    private ICache<String, IpAddrRange> ipAddrRangeCache;
+    private ICache<String, VpcIpRange> vpcIpRangeCache;
+    private Transaction transaction;
+
+    @BeforeEach
+    public void beforeEachTest() throws Exception {
+        transaction = mock(Transaction.class);
+        ipAddrRangeCache = spy(new MockCache<>(transaction));
+        vpcIpRangeCache = spy(new MockCache<>(transaction));
+
+        ipAddrRangeRepo = new IpAddrRangeRepo(ipAddrRangeCache, vpcIpRangeCache);
+
+        when(transaction.start()).thenReturn(transaction);
+
+        IpAddrRangeRequest ipAddrRangeRequest = buildIpAddrRangeRequest();
+        ipAddrRangeRepo.createIpAddrRange(ipAddrRangeRequest);
+    }
+
+    @Test
+    public void allocateIpAddrWithIpAddrTest() throws Exception {
+        IpAddrRequest ipAddrRequest = buildIpAddrRequest();
+        IpAddrAlloc ipAddrAlloc = ipAddrRangeRepo.allocateIpAddr(ipAddrRequest);
+        assertNotNull(ipAddrAlloc);
+        assertEquals(ipAddrAlloc.getIpAddr(), UnitTestConfig.ip1);
+        assertEquals(ipAddrAlloc.getState(), UnitTestConfig.activated);
+    }
+
+    @Test
+    public void allocateIpAddrWithoutIpAddrTest() throws Exception {
+        IpAddrRequest ipAddrRequest = buildIpAddrRequest();
+        ipAddrRequest.setIp(null);
+        IpAddrAlloc ipAddrAlloc = ipAddrRangeRepo.allocateIpAddr(ipAddrRequest);
+        assertNotNull(ipAddrAlloc);
+        assertEquals(ipAddrAlloc.getIpAddr(), UnitTestConfig.ip1);
```

Comment:
This assertion shows that our algorithm is predictable :-)

Reply:
yeah, sequentially.